source/common/upstream/deterministic_aperture_lb.cc

```diff
@@ -0,0 +1,150 @@
+  return {ALL_DETERMINISTIC_APERTURE_LOAD_BALANCER_RING_STATS(POOL_COUNTER(scope))};
+}
+
+/*
+ * TODO(jojy): The concept of HashingLoadBalancer::chooseHost might not directly apply here since we
+ * don't actually use the hash of the nodes placed in the ring. We use `random` algorithm in the
+ * Deterministic aperture load balancer. Hence we ignore the `h` and `attempt` here.
+ */
+HostConstSharedPtr DeterministicApertureLoadBalancer::Ring::chooseHost(uint64_t, uint32_t) const {
+  if (ring_.empty()) {
+    return nullptr;
+  }
+
+  const std::pair<size_t, size_t> index_pair = pick2();
+
+  const RingHashLoadBalancer::RingEntry& first = ring_[index_pair.first];
+  const RingHashLoadBalancer::RingEntry& second = ring_[index_pair.second];
+
+  ENVOY_LOG(debug, "pick2 returned hosts: (hash1: {}, address1: {}, hash2: {}, address2: {})",
+            first.hash_, first.host_->address()->asString(), second.hash_,
+            second.host_->address()->asString());
+
+  return first.host_->stats().rq_active_.value() < second.host_->stats().rq_active_.value()
+             ? first.host_
+             : second.host_;
+}
+
+using HashFunction = envoy::config::cluster::v3::Cluster::RingHashLbConfig::HashFunction;
+DeterministicApertureLoadBalancer::Ring::Ring(
+    double offset, double width, const NormalizedHostWeightVector& normalized_host_weights,
+    double min_normalized_weight, uint64_t min_ring_size, uint64_t max_ring_size,
+    HashFunction hash_function, bool use_hostname_for_hashing, Stats::ScopeSharedPtr scope,
+    RingHashLoadBalancerStats ring_stats)
+    : RingHashLoadBalancer::Ring(normalized_host_weights, min_normalized_weight, min_ring_size,
+                                 max_ring_size, hash_function, use_hostname_for_hashing,
+                                 ring_stats),
+      offset_(offset), width_(width), unit_width_(1.0 / ring_size_), rng_(random_dev_()),
+      random_distribution_(0, 1), stats_(generateStats(*scope)) {
+  if (width_ > 1.0 || width_ < 0) {
+    throw EnvoyException(
+        fmt::format("Invalid width for the deterministic aperture ring{}", width_));
+  }
+}
+
+absl::optional<double> DeterministicApertureLoadBalancer::Ring::weight(size_t index, double offset,
+                                                                       double width) const {
+  if (index >= ring_size_ || width > 1 || offset > 1) {
+    return absl::nullopt;
+  }
+
+  double index_begin = index * unit_width_;
+  double index_end = index_begin + unit_width_;
+
+  if (offset + width > 1) {
+    double start = std::fmod((offset + width), 1.0);
+
+    return 1.0 - (intersect(index_begin, index_end, start, offset)) / unit_width_;
+  }
+
+  return intersect(index_begin, index_end, offset, offset + width) / unit_width_;
+}
+
+size_t DeterministicApertureLoadBalancer::Ring::getIndex(double offset) const {
+  RELEASE_ASSERT(offset >= 0 && offset <= 1, "valid offset");
```

Comment:
I think RELEASE_ASSERT is probably a little aggressive. We normally reserve it for cases where Envoy cannot continue: out of memory, for instance.

My opinion is that we should use a normal ASSERT here and force offset into the expected range if it's out of bounds due to some corner case of rounding or floating point precision error.